strawberryfields/backends/bosonicbackend/backend.py

```diff
@@ -54,6 +69,205 @@ def __init__(self):
         self._supported["mixed_states"] = True
         self._init_modes = None
         self.circuit = None
+        self.ancillae_samples_dict = {}
+
+    def run_prog(self, prog, **kwargs):
```

Comment:
How come this method is defined? Mostly curious about how it is going to be used with the engine as other backends do not seem to define such a method.

Reply:
@antalszava : the idea here is that we want to have a custom `run_prog` that checks the type of initialization that happens for the different modes. If you know in advance how the different modes are initialized then you can directly allocate the required arrays. Otherwise, everytime a new mode is parsed one would have to generate a new array with the extra dimensions required for the particular preparation. For example if you have a circuit with two cats, this method will recognize that it needs 4x4 = 16 coefficients, means and covariances. If you instead used the usual `run_prog` you would first allocate 4 coefficients, means and covariances and then later when you read the preparation of the second cat you would have to generate the new arrays having a first index running from 0 to 16-1.

Reply:
It was decided early on to build a custom engine for the bosonic backend, which is essentially what this method is, mainly since we had some complications with the initializations for each mode. You will see the changes made to the engine file in the next PR. I'm tagging @ilan-tz since he can give a more thorough answer.

Reply:
Yep. The original backends (Fock, Gaussian, and TF) all are run 'step-by-step'. The `LocalEngine` will initialize the backend with the number of modes and some other program independent properties, and will then separately loop over the program operations and call the corresponding backend methods.

This does not work well for the bosonic backend; information about the program is needed at initialization, but the current `LocalEngine` does not provide! In fact, the `LocalEngine` tries to do too much by itself.

A better approach is how the `RemoteEngine` works - it passes the program _directly_ to the backend, waits for the backend to interpret and execute it, and receives the results. This is the approach we have decided to go with for the Bosonic backend.

Reply:
Sidenote: we did look into attempting to modify the local engine to provide this information to the backend, but you end up going down a rabbit hole of needing to patch the local engine, patch the _existing_ backends, and still have the whole thing look like a bit of a hotfix.